src/proto/h2/bdp.rs

```diff
@@ -0,0 +1,190 @@
+//
+// # BDP Algorithm
+//
+// 1. When receiving a DATA frame, if a BDP ping isn't outstanding:
+//   1a. Record current time.
+//   1b. Send a BDP ping.
+// 2. Increment the number of received bytes.
+// 3. When the BDP ping ack is received:
+//   3a. Record duration from sent time.
+//   3b. Merge RTT with a running average.
+//   3c. Calculate bdp as bytes/rtt.
+//   3d. If bdp is over 2/3 max, set new max to bdp and update windows.
+//
+//
+// # Implementation
+//
+// - `hyper::Body::h2` variant includes a "bdp channel"
+//   - When the body's `poll_data` yields bytes, call `bdp.sample(bytes.len())`
+//
+
+use std::sync::{Arc, Mutex, Weak};
+use std::task::{self, Poll};
+use std::time::{Duration, Instant};
+
+use h2::{Ping, PingPong};
+
+type WindowSize = u32;
+
+/// Any higher than this likely will be hitting the TCP flow control.
+const BDP_LIMIT: usize = 1024 * 1024 * 16;
+
+pub(crate) fn disabled() -> Sampler {
+    Sampler {
+        shared: Weak::new(),
+    }
+}
+
+pub(super) fn channel(ping_pong: PingPong, initial_window: WindowSize) -> (Sampler, Estimator) {
+    let shared = Arc::new(Mutex::new(Shared {
+        bytes: 0,
+        ping_pong,
+        ping_sent: false,
+        sent_at: Instant::now(),
+    }));
+
+    (
+        Sampler {
+            shared: Arc::downgrade(&shared),
+        },
+        Estimator {
+            bdp: initial_window,
+            max_bandwidth: 0.0,
+            shared,
+            samples: 0,
+            rtt: 0.0,
+        },
+    )
+}
+
+#[derive(Clone)]
+pub(crate) struct Sampler {
+    shared: Weak<Mutex<Shared>>,
+}
+
+pub(super) struct Estimator {
+    shared: Arc<Mutex<Shared>>,
+
+    /// Current BDP in bytes
+    bdp: u32,
+    /// Largest bandwidth we've seen so far.
+    max_bandwidth: f64,
+    /// Count of samples made (ping sent and received)
+    samples: usize,
+    /// Round trip time in seconds
+    rtt: f64,
+}
+
+struct Shared {
+    bytes: usize,
+    ping_pong: PingPong,
+    ping_sent: bool,
+    sent_at: Instant,
+}
+
+impl Sampler {
+    pub(crate) fn sample(&self, bytes: usize) {
+        let shared = if let Some(shared) = self.shared.upgrade() {
+            shared
+        } else {
+            return;
+        };
+
+        let mut inner = shared.lock().unwrap();
+
+        if !inner.ping_sent {
+            if let Ok(()) = inner.ping_pong.send_ping(Ping::opaque()) {
+                inner.ping_sent = true;
+                inner.sent_at = Instant::now();
+                trace!("sending BDP ping");
+            } else {
+                return;
+            }
+        }
+
+        inner.bytes += bytes;
+    }
+
+    pub(crate) fn is_enabled(&self) -> bool {
+        self.shared.strong_count() > 0
+    }
+}
+
+impl Estimator {
+    pub(super) fn poll_estimate(&mut self, cx: &mut task::Context<'_>) -> Poll<WindowSize> {
+        let mut inner = self.shared.lock().unwrap();
+        if !inner.ping_sent {
+            // XXX: this doesn't register a waker...?
```

Comment:
Well it shouldn't need to, since sending the ping will wake up the h2 connection task, which is the same one... It's just that `poll_estimate` is assuming that even though the function doesn't otherwise know.